tests/sycl/group_functions.cpp

```diff
@@ -0,0 +1,1800 @@
+/*
+ * This file is part of hipSYCL, a SYCL implementation based on CUDA/HIP
+ *
+ * Copyright (c) 2018-2020 Aksel Alpay and contributors
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#include "sycl_test_suite.hpp"
+#include <cstddef>
+#include <cstdio>
+#include <functional>
+#include <hipSYCL/sycl/libkernel/host/host_backend.hpp>
+#include <iostream>
+#include <limits>
+#include <math.h>
+#include <type_traits>
+
+#include <string>
+#include <sstream>
+
+using namespace cl;
+
+#ifdef TESTS_GROUPFUNCTION_FULL
+using test_types = boost::mpl::list<char, int, unsigned int,long long, float, double, sycl::vec<int, 1>, sycl::vec<int, 2>, sycl::vec<int, 3>, sycl::vec<int, 4>, sycl::vec<int, 8>,  sycl::vec<short, 16>, sycl::vec<long, 3>, sycl::vec<unsigned int, 3>>;
+#else
+using test_types = boost::mpl::list<char, unsigned int, float, double, sycl::vec<int, 2>>;
+#endif
+
+namespace detail {
+
+template<typename T>
+using elementType = typename std::remove_reference<decltype(T{}.s0())>::type;
+
+template<typename T, int N>
+std::string T2str(sycl::vec<T, N> v)
+{
+    std::stringstream ss{};
```

Comment:
We use 2 spaces as indentation (`clang-format` can do it automatically for you  :-) )